### PR TITLE
Problem (fix #269): denom conversion in cosmos-sdk is not used

### LINKED ISSUE #269: Problem: denom conversion in cosmos-sdk is not used
https://github.com/cosmos/cosmos-sdk/pull/7777

denom conversion and decimal support is merged in cosmos-sdk now, we can cleanup our forked cli command code now.

### PR DESCRIPTION
**Pull request by yihuang**
Solution:
- update to cosmos-sdk master
- use sdk.ParseCoinsNormalized
- cleanup unused code
- re-organize code to minimize diff between app/app.go and
  simapp/app.go, cmd/chain-maind/app/app.go and simapp/simd/cmd/root.go
- disable line length linter(lll), to keep it consistent with cosmos-sdk coding style
- one can still use the derivation of mkShell to work with remote shell

👮🏻👮🏻👮🏻 !!!! REFERENCE THE PROBLEM YOUR ARE SOLVING IN THE PR TITLE AND DESCRIBE YOUR SOLUTION HERE !!!! DO NOT FORGET !!!! 👮🏻👮🏻👮🏻

# PR Checklist:

- [x] Have you read the [CONTRIBUTING.md](https://github.com/crypto-com/chain-main/blob/master/CONTRIBUTING.md)?
- [x] Does your PR follow the [C4 patch requirements](https://rfc.zeromq.org/spec:42/C4/#23-patch-requirements)?
- [x] Have you rebased your work on top of the latest master? 
- [x] Have you checked your code compiles? (`make`)
- [x] Have you included tests for any non-trivial functionality?
- [x] Have you checked your code passes the unit tests? (`make test`)
- [x] Have you checked your code formatting is correct? (`go fmt`)
- [x] Have you checked your basic code style is fine? (`golangci-lint run`)
- [x] If you added any dependencies, have you checked they do not contain any known vulnerabilities? (`go list -json -m all | nancy sleuth`)
- [x] If your changes affect the client infrastructure, have you run the integration test?
- [x] If your changes affect public APIs, does your PR follow the [C4 evolution of public contracts](https://rfc.zeromq.org/spec:42/C4/#26-evolution-of-public-contracts)?
- [x] If your code changes public APIs, have you incremented the crate version numbers and documented your changes in the [CHANGELOG.md](https://github.com/crypto-com/chain-main/blob/master/CHANGELOG.md)?
- [x] If you are contributing for the first time, please read the agreement in [CONTRIBUTING.md](https://github.com/crypto-com/chain-main/blob/master/CONTRIBUTING.md) now and add a comment to this pull request stating that your PR is in accordance with the [Developer's Certificate of Origin](https://github.com/crypto-com/chain-main/blob/master/CONTRIBUTING.md#developer-certificate-of-originn).

Thank you for your code, it's appreciated! :)

### CODE REVIEW

**Comment by yihuang** [comment]
it seems the deadcode linter didn't work in our project.

File: go.mod [hunk 1]
```diff
@@ -4,26 +4,26 @@ go 1.15
...
+	google.golang.org/genproto v0.0.0-20201111145450-ac7456db90a6
+	google.golang.org/grpc v1.33.2
 
 )
 
 replace github.com/gogo/protobuf => github.com/regen-network/protobuf v1.3.2-alpha.regen.4
 
 replace github.com/cosmos/ledger-cosmos-go => github.com/crypto-com/ledger-cosmos-go v0.9.10-0.20200929055312-01e1d341de0f
+
+replace github.com/cosmos/cosmos-sdk => github.com/yihuang/cosmos-sdk v0.34.4-0.20201117081135-e71014e005c0
```
Comment (tomtau) [inline go.mod:29]: what is this custom fork -- https://github.com/yihuang/cosmos-sdk/commit/e71014e005c0 ? we'll ideally use the released versions.
the coin parsing may not make it to a release.
I can potentially cherry-pick the necessary commits over "v0.40.*" or "release/v0.40" (when it exists), but it'll ideally be some finalised commit that made it to master

**Review by tomtau — COMMENTED** [review]
looks good -- it'll reduce a lot of unnecessary boilerplate/module wrapping.
I'd wait though for a more finalised version

File: go.mod (hunk 1 above)
Comment (yihuang) [reply go.mod:29]: the changed ParseCoins is already merged into master, but I find the ParseCoin also need to change, hence a second PR. We can wait for it to be merged.

File: pystarport/pystarport/cluster.py [hunk 2]
```diff
@@ -400,7 +401,8 @@ def collect_gentxs(self, gentx_dir, i=0):
         return self.raw("collect-gentxs", gentx_dir, home=self.home(i))
 
     def status(self, i=0):
-        return json.loads(self.raw("status", node=self.node_rpc(i)))
+        url = "http" + self.node_rpc(i)[3:]
```
Comment (yihuang) [inline pystarport/pystarport/cluster.py:404]: This is a workaround for bug: https://github.com/cosmos/cosmos-sdk/issues/7980

Comment (yihuang) [reply pystarport/pystarport/cluster.py]: this bug is fixed upstream, revert it back.

File: go.mod (hunk 1 above)
Comment (yihuang) [reply go.mod]: upstream pr is merged now.

File: integration_tests/test_grpc.py [hunk 3]
```diff
@@ -1,11 +1,15 @@
...
-from pystarport.proto_python.grpc_util import GrpcUtil, GrpcUtilTxBroadcast
+import pytest
 
 from .utils import wait_for_new_blocks
 
+pytestmark = pytest.mark.grpc
+
 
 def test_query_validators(cluster):
+    from pystarport.proto_python.grpc_util import GrpcUtil
```
Comment (yihuang) [inline integration_tests/test_grpc.py:11]: move it here so when proto files not prepared locally, it don't prevent other tests from running.

File: integration_tests/utils.py [hunk 4]
```diff
@@ -19,26 +19,26 @@ def wait_for_block(cli, height, timeout=60):
         except AssertionError as e:
             print(f"get sync status failed: {e}", file=sys.stderr)
         else:
-            if int(status["sync_info"]["latest_block_height"]) >= height:
+            if int(status["SyncInfo"]["latest_block_height"]) >= height:
```
Comment (leejw51crypto) [inline integration_tests/utils.py:22]: why changed to camel style?

File: x/chainmain/client/cli/query.go [hunk 5]
```diff
@@ -74,11 +74,18 @@ Example:
 				return allBalancesErr
 			}
 
-			baseUnit := coinParser.GetBaseUnit()
+			baseUnit, err := sdk.GetBaseDenom()
+			if err != nil {
+				return err
+			}
 			baseAmount := res.Balances.AmountOf(baseUnit)
+			humanCoin, err := sdk.ConvertDecCoin(sdk.NewDecCoin(baseUnit, baseAmount), config.HumanCoinUnit)
```
Comment (leejw51crypto) [inline x/chainmain/client/cli/query.go:82]: comment about ConvertDecCoin would be good

**Review by leejw51crypto — APPROVED** [review]
lgtm

File: integration_tests/utils.py (hunk 4 above)
Comment (yihuang) [reply integration_tests/utils.py:22]: it's changed in cosmos-sdk in this PR: https://github.com/cosmos/cosmos-sdk/pull/7974/files

File: x/chainmain/client/cli/query.go (hunk 5 above)
Comment (yihuang) [reply x/chainmain/client/cli/query.go:82]: It converts the coin to the specified denomination, for example `"1000000basecro" -> "1cro"`

**Comment by yihuang** [comment]
it seems relayer don't work yet after migrate to master cosmos-sdk.

**Comment by yihuang** [comment]
I've cherry-picked denom-conversion commits on top of rc3 here: https://github.com/yihuang/cosmos-sdk/tree/rc3

**Comment by tomtau** [comment]
denom conversion is now in rc5 -- https://github.com/crypto-com/chain-main/pull/297

**Pull request closed without merging**